compiler/src/iree/compiler/Dialect/HAL/Utils/ExternBuildingUtils.cpp

```diff
@@ -0,0 +1,91 @@
+#include "iree/compiler/Dialect/HAL/IR/HALDialect.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "mlir/Dialect/PDLInterp/IR/PDLInterp.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/IR/Location.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/RegionUtils.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace HAL {
+
+namespace {
+
+/// Helper to build a hal.dispatch.extern op with the given arguments. This
+/// returns the block arguments of the workgroup count region and the extern
+/// op. Note that the workgroup count region will not include the terminator
+/// and that is left up to the user to properly populate.
+static FailureOr<Operation *>
+createDispatchExtern(PatternRewriter &rewriter, ValueRange workload,
+                     TypeRange resultTypes, ValueRange resultDims,
+                     ValueRange arguments, ValueRange argumentDims,
+                     DenseI64ArrayAttr tiedOperands, DictionaryAttr attrDict) {
+  Location rootLoc = (*arguments.begin()).getLoc();
+  SmallVector<int64_t> tiedOperandsIntList(tiedOperands.asArrayRef());
+  SmallVector<NamedAttribute> namedAttributes(attrDict.begin(), attrDict.end());
+  Operation *externOp = rewriter.create<DispatchExternOp>(
+      rootLoc, workload, resultTypes, resultDims, arguments, argumentDims,
+      tiedOperandsIntList, namedAttributes);
+  return externOp;
+}
+
+/// Helper to emplace a block on the given hal.dispatch.extern op. This returns
+/// the block arguments of the updated workgroup count region. Note that the
+/// workgroup count region will not include the terminator and that is left up
+/// to the user to properly populate.
+static FailureOr<ValueRange>
+emplaceExternWorkgroupCountRegion(PatternRewriter &rewriter, Operation *op) {
+  Location rootLoc = op->getLoc();
+  auto externOp = dyn_cast<DispatchExternOp>(op);
+  if (!externOp) {
+    return failure();
+  }
+
+  SmallVector<Type> countTypes({rewriter.getType<IREE::HAL::DeviceType>()});
+  SmallVector<Location> countLocs({rootLoc});
+  for (auto workloadIdx : externOp.getWorkload()) {
+    countTypes.push_back(workloadIdx.getType());
+    countLocs.push_back(workloadIdx.getLoc());
+  }
+
+  auto &entryBlock = externOp.getWorkgroupCount().emplaceBlock();
+  auto countArgs = entryBlock.addArguments(countTypes, countLocs);
+
+  ArrayRef<BlockArgument> countArgsArray(countArgs.begin(), countArgs.end());
+
+  /// Update the insertion point to the beginning of the block to enable
+  /// contructing the workgroup count region.
+  rewriter.setInsertionPointToStart(&entryBlock);
+  return ValueRange(countArgsArray);
+}
+} // namespace
+
+void registerExternDispatchRewriteFunction(PDLPatternModule &pdlPatterns) {
+  pdlPatterns.registerRewriteFunction("create_dispatch_extern",
```

Comment:
these can be moved into MaterializeDispatchExterns - keeps them local to where they are used - and once the two methods are on the op take lambda fns to wrap for less code:

```
 pdlPatterns.registerRewriteFunction("create_dispatch_extern", [](PatternRewriter &rewriter, ValueRange workload,
                     TypeRange resultTypes, ValueRange resultDims,
                     ValueRange arguments, ValueRange argumentDims,
                     DenseI64ArrayAttr tiedOperands, DictionaryAttr attrDict) {
  return rewriter.create<IREE::HAL::DispatchExternOp>(....);
});
```

Reply:
I've had enough hard to understand crashes with PDL thus far (including some related to the fact that the native rewrite ops aren't verified until interpretation time) that I think leaving the helpers as their own function is worth it as it helps make the stack trace a little more readable.